src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,429 @@
+    }
+
+    /// Calculates the API key ID of an `Event`
+    ///
+    /// This function calculates the API key ID of a given `Event`. As a
+    /// side-effect it mutates internal state of the struct allowing callers to
+    /// use the ID to retrieve a `Box<str>` of the key at a later time.
+    fn register_key_id(&mut self, event: &Event) -> u64 {
+        if let Some(api_key) = event.metadata().datadog_api_key() {
+            let key = api_key.as_ref();
+            let key_hash = hash(key);
+            // TODO it'd be nice to avoid passing through String
+            self.key_slab
+                .entry(key_hash)
+                .or_insert_with(|| String::from(key).into_boxed_str());
+            key_hash
+        } else {
+            self.default_api_key
+        }
+    }
+
+    fn has_space(&self, id: u64) -> bool {
+        if let Some(arr) = self.event_batches.get(&id) {
+            arr.len() < MAX_PAYLOAD_ARRAY
+        } else {
+            true
+        }
+    }
+
+    /// Stores the `Event` in its batch
+    ///
+    /// This function stores the `Event` in its `id` appropriate batch. Caller
+    /// MUST confirm that there is space in the batch for the `Event` by calling
+    /// `has_space`, which must return true.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is no space in the underlying batch.
+    fn store_event(&mut self, id: u64, event: Event) {
+        let arr = self
+            .event_batches
+            .entry(id)
+            .or_insert_with(|| Vec::with_capacity(MAX_PAYLOAD_ARRAY));
+        assert!(arr.len() + 1 <= MAX_PAYLOAD_ARRAY);
+        arr.push(event);
+    }
+
+    /// Flush the `Event` batches out to Datadog API
+    async fn flush_to_api<'a>(
+        self: &'a mut Self,
+    ) -> Result<impl Iterator<Item = Client::Future> + Send + 'a, ()> {
+        // The Datadog API makes no claim on how many requests we can make at
+        // one time. We rely on the http client to handle retries and what
+        // not. This implementation stacks up futures for each key present in
+        // the batches and runs them unordered, returning when the last
+        // completes.
+
+        // NOTE I would prefer to map directly over the Drain without going
+        // through a vec but mixing ownership confused me greatly, since we need
+        // to call back to `self` to flush the batch.
+        let batches: Vec<(u64, Vec<Event>)> = self.event_batches.drain().collect();
+        poll_fn(|cx| self.http_client.poll_ready(cx))
+            .await
+            .map_err(|_e| ())?;
+        let futures = // : dyn Iterator<Item = Client::Future> =
+            batches.into_iter().map(move |(api_key_id, batch)| {
+                let members: Vec<BTreeMap<String, Value>> = batch
+                    .into_iter()
+                    .map(|event| event.into_log())
+                    .map(|log| log.into_parts().0)
+                    .collect();
```

Comment:
Ok, so here you break the log events into parts, and drop the metadata. This marks the event as "dropped" for finalization. I think your best bet here is to iterate through the batch and map them with `event.as_mut_log().take_finalizer()` and collect those into a batch. ie:
```rust
let finalizers: Vec<_> = batch.iter_mut().map(|event| event.as_mut_log().take_finalizer()).collect();
```